user.go

```diff
@@ -110,107 +110,141 @@ func (c *Client) UnmuteUsers(targetIDs []string, userID string) error {
 	return c.makeRequest(http.MethodPost, "moderation/unmute", nil, data, nil)
 }
 
-func (c *Client) FlagUser(targetID string, options map[string]interface{}) error {
+func (c *Client) FlagUser(targetID string, options ...Option) error {
 	switch {
 	case targetID == "":
-		return errors.New("target ID is empty")
+		return ErrorMissingTargetID
 	case len(options) == 0:
 		return errors.New("flag user: options must be not empty")
 	}
 
-	options["target_user_id"] = targetID
+	options = append(options, NewOption(optionKeyTargetUserID, targetID))
 
-	return c.makeRequest(http.MethodPost, "moderation/flag", nil, options, nil)
+	return c.makeRequestWithOptions(http.MethodPost, "moderation/flag", nil, options, nil)
 }
 
-func (c *Client) UnFlagUser(targetID string, options map[string]interface{}) error {
-	switch {
-	case targetID == "":
-		return errors.New("target ID is empty")
-	case options == nil:
-		options = map[string]interface{}{}
+func (c *Client) UnFlagUser(targetID string, options ...Option) error {
+	if targetID == "" {
+		return ErrorMissingTargetID
 	}
 
-	options["target_user_id"] = targetID
+	options = append(options, NewOption(optionKeyTargetUserID, targetID))
 
-	return c.makeRequest(http.MethodPost, "moderation/unflag", nil, options, nil)
+	return c.makeRequestWithOptions(http.MethodPost, "moderation/unflag", nil, options, nil)
 }
 
-func (c *Client) BanUser(targetID, userID string, options map[string]interface{}) error {
-	switch {
-	case targetID == "":
-		return errors.New("target ID is empty")
-	case userID == "":
-		return errors.New("user ID is empty")
-	case options == nil:
-		options = map[string]interface{}{}
-	}
+func (c *Client) BanUser(targetID, userID string, options ...Option) error {
+	return c.banUser(&banUserInput{
+		TargetID: targetID,
+		UserID:   userID,
+	}, options...)
+}
+
+type banUserInput struct {
+	// TargetID is the ID of the person to be banned.
+	TargetID string `json:"target_user_id,omitempty"`
 
-	options["target_user_id"] = targetID
-	options["user_id"] = userID
+	// UserID is the ID of the person doing the banning.
+	UserID string `json:"user_id,omitempty"`
 
-	return c.makeRequest(http.MethodPost, "moderation/ban", nil, options, nil)
+	Reason  *string `json:"reason,omitempty"`
+	Timeout *int    `json:"timeout,omitempty"`
+
+	// ChannelType and ChannelID are used if the ban is channel specific.
+	ChannelType *string `json:"type,omitempty"`
+	ChannelID   *string `json:"id,omitempty"`
+
+	Extra map[string]interface{} `json:"-"`
 }
 
-func (c *Client) UnBanUser(targetID string, options map[string]string) error {
+func (b *banUserInput) AddOptions(options []Option) {
+	for _, opt := range options {
+		switch opt.Key() {
+		case optionKeyTimeout:
+			seconds := opt.Value().(int)
+			b.Timeout = &seconds
+		case optionKeyReason:
+			reason := optionAsString(opt)
+			b.Reason = &reason
+		}
```

Comment:
not sure if we should fail on unknown keys to let users that they are doing something wrong